sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,180 @@
+    assert PartR.masscenter.vel(A) == l*u2*B.y
+
+def test_pin_joint_chaos_pendulum():
+    mA, mB, lA, lB, h = symbols('mA, mB, lA, lB, h')
+    theta, phi, omega, alpha = dynamicsymbols('theta phi omega alpha')
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    B = ReferenceFrame('B')
+    lA = (lB - h / 2) / 2
+    lC = (lB/2 + h/4)
+    rod = Body('rod', frame=A, mass=mA)
+    plate = Body('plate', mass=mB, frame=B)
+    C = Body('C', frame=N)
+    J1 = PinJoint('J1', C, rod, coordinates=theta, speeds=omega, child_joint_pos=lA*A.z,\
+        parent_axis=N.y, child_axis=A.y)
+    J2 = PinJoint('J2',rod, plate, coordinates=phi, speeds=alpha, parent_joint_pos=lC*A.z,\
+        parent_axis=A.z, child_axis=B.z)
+
+    #Check orientation
+    assert A.dcm(N) == Matrix([[cos(theta), 0, -sin(theta)], [0, 1, 0], [sin(theta), 0, cos(theta)]])
+    assert A.dcm(B) == Matrix([[cos(phi), -sin(phi), 0], [sin(phi), cos(phi), 0], [0, 0, 1]])
+    assert B.dcm(N) == Matrix([[cos(phi)*cos(theta), sin(phi), -sin(theta)*cos(phi)], \
+        [-sin(phi)*cos(theta), cos(phi), sin(phi)*sin(theta)], [sin(theta), 0, cos(theta)]])
+
+    #Check Angular Velocity
+    assert A.ang_vel_in(N) == omega * N.y
+    assert A.ang_vel_in(B) == - alpha * A.z
+    assert N.ang_vel_in(B) == - omega * N.y - alpha * A.z
+
+    #Check kde
+    assert J1.kdes() == [omega - theta.diff(t)]
+    assert J2.kdes() == [alpha - phi.diff(t)]
+
+    #Check pos of masscenters
+    assert C.masscenter.pos_from(rod.masscenter) == lA*A.z
+    assert rod.masscenter.pos_from(plate.masscenter) == - lC * A.z
+
+    #Check Linear Velocities
+    assert rod.masscenter.vel(N) == (h/4 - lB/2)*omega*A.x
+    assert plate.masscenter.vel(N) == ((h/4 - lB/2)*omega + (h/4 + lB/2)*omega)*A.x
+
+def test_pinjoint_arbitrary_axis():
+    def generate_body():
+        N = ReferenceFrame('N')
+        A = ReferenceFrame('A')
+        P = Body('P', frame=N)
+        C = Body('C', frame=A)
+        return N, A, P, C
+
+    theta, omega = dynamicsymbols('J_theta, J_omega')
+
+    #When the bodies are attached though masscenters but axess are opposite.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == 0
+    assert A.dcm(N) == Matrix([[1, 0, 0], [0, cos(theta), sin(theta)], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert C.masscenter.pos_from(P.masscenter) == 0
+    assert C.masscenter.vel(N) == 0
+
+    #When axes are different and parent joint is at masscenter but child joint is at a unit vector from
+    #child masscenter.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=A.y, child_joint_pos=A.x)
+
+    assert A.y.angle_between(N.x) == 0 #Axis are aligned
+    assert A.dcm(N) == Matrix([[0, -cos(theta), -sin(theta)], [1, 0, 0], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert C.masscenter.vel(N) == omega*A.z
```

Comment:
If you make the child_joint_pos something other than a unit vector, we'd see that this is fully computed correctly. Make it `v*A.x` for example.